(nan)quantile CPU optimization

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44497 (nan)quantile CPU optimization**
* #44393 Fixed quantile nan propagation and implemented nanquantile

This PR replaces a call to sort with a multi-quickselect implementation to partition the data before querying for kth values. When the number of quantiles is small compared to the size of the dimension being reduced, this leads to big performance improvements.

For a single quantile, the previous code was 8 times slower than NumPy. Now it is the 1-2x faster for quantile and up to 8x faster for nanquantile. The benchmark snippet below demonstrates this case.

```
import torch
import numpy as np
import pytest

torch.set_num_threads(1)
torch.set_grad_enabled(False)

x = torch.randn(1000, 1000, dtype=torch.float64)
q = torch.rand(1, dtype=torch.float64)


@pytest.mark.parametrize('op', ['torch', 'numpy'])
def test_quantile(benchmark, op):
    if op == 'torch':
        benchmark(torch.nanquantile, x, q, 1)
    else:
        benchmark(np.nanquantile, x.numpy(), q.numpy(), 1)
```

### Outputs
```
---------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------
Name (time in ms)            Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_quantile[torch]     10.8364 (1.0)      13.4978 (1.0)      11.6419 (1.0)      0.5855 (1.0)      11.7781 (1.0)      0.9811 (1.92)         27;1  85.8970 (1.0)          67           1
test_quantile[numpy]     94.6725 (8.74)     96.9456 (7.18)     95.4756 (8.20)     0.6195 (1.06)     95.3307 (8.09)     0.5101 (1.0)           3;1  10.4739 (0.12)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

heitorschueroff committed:
(nan)quantile CPU optimization

[ghstack-poisoned]

heitorschueroff committed:
Update on "(nan)quantile CPU optimization"




[ghstack-poisoned]

heitorschueroff committed:
Update on "(nan)quantile CPU optimization"


This PR replaces a call to sort with a multi-quickselect implementation to partition the data before querying for kth values. When the number of quantiles is small compared to the size of the dimension being reduced, this leads to big performance improvements.

For a single quantile, the previous code was 8 times slower than NumPy. Now it is the 1-2x faster for quantile and up to 8x faster for nanquantile. The benchmark snippet below demonstrates this case.

```
import torch
import numpy as np
import pytest

torch.set_num_threads(1)
torch.set_grad_enabled(False)

x = torch.randn(1000, 1000, dtype=torch.float64)
q = torch.rand(1, dtype=torch.float64)


@pytest.mark.parametrize('op', ['torch', 'numpy'])
def test_quantile(benchmark, op):
    if op == 'torch':
        benchmark(torch.nanquantile, x, q, 1)
    else:
        benchmark(np.nanquantile, x.numpy(), q.numpy(), 1)
```

### Outputs
```
---------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------
Name (time in ms)            Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_quantile[torch]     10.8364 (1.0)      13.4978 (1.0)      11.6419 (1.0)      0.5855 (1.0)      11.7781 (1.0)      0.9811 (1.92)         27;1  85.8970 (1.0)          67           1
test_quantile[numpy]     94.6725 (8.74)     96.9456 (7.18)     95.4756 (8.20)     0.6195 (1.06)     95.3307 (8.09)     0.5101 (1.0)           3;1  10.4739 (0.12)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

[ghstack-poisoned]

mruberry commented:
These numbers look really good. What if the number of threads isn't constrained and a variety of sizes are used? Let's say, 1D small medium and large, 2D small medium and large, where small=128, medium=32x512, large = 1024x1024?

So the performance cross we want to look at is [1 thread, unconstrained threads] x [1D small medium large, 2D small medium large] x [quantile, nanquantile] x [torch, numpy].

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
 
+// Returns pointer to pivot used to partition range [first, last]
+// There should be at least 3 elements in [first, last]
+template <typename random_it, typename comp>
```
Do you want to add a brief description of the algorithm you're using, too?

mruberry replied:
Also: what are the invariants the arguments to this function need to satisfy, and what are the invariants coming out of it?

mruberry replied:
Second question: how does this algorithm compare to std::partition?

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+template <typename random_it, typename comp>
+random_it partition(random_it first, random_it last, comp lt) {
+  // Use median of three for pivot choice
+  random_it pivot = first + (std::distance(first, last) >> 1);
```
To improve readability probably just want to divide by 2. The compiler should figure out that it can do a shift (or not).

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+random_it partition(random_it first, random_it last, comp lt) {
+  // Use median of three for pivot choice
+  random_it pivot = first + (std::distance(first, last) >> 1);
+  std::swap(first[1], *pivot);
```
Would you add a comment here describing what this block does (what invariant is it setting up?)

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+}
+
+template <typename random_it, typename kth_iterator, typename comp>
+void quickselect_helper(
```
Plus a comment describing this function, too

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+    }
+    return;
+  }
+
```
To be parallel with previous comments, this would have a comment like "General case: 3 or more elements"

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+  quickselect_helper(start, pivot + 1, last, gt_kbegin, kend, lt);
+}
+
+// Rearranges elements in range [first, last) such that kths elements
```
"kths elements" is a little vague. Can you elaborate on what the input "kths" should be / what it represents and what the property this algorithm achieves? Maybe a brief sentence with an example would help. "For example, if "kths" contains 2, 5, and 7, then ..."

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+      first, first, last - 1, kths.begin(), kths.end(), lt);
+}
+
+void quantile_cpu_impl(
```
Plus a comment here briefly introducing the function, then explaining the algorithm, then adding a brief "NOTE: ..." explaining the CPU (nan)quantile implementation is different than other backends because it's been optimized for performance.

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+      q.ge(0).logical_and_(q.le(1)).all().item<bool>(),
+      "quantile() q values must be in the range [0, 1]");
+
+  // If dim is None flatten. Move dimension to reduce as last dim and flatten
```
This could use a slight tweak. Something like... "If dim is None then quantiles are computed on the flattened tensor. If dim is not None, ..."?

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+  return last;
+}
+
+template <typename random_it, typename kth_iterator, typename comp>
```
naming the comparator "lt" is pretty suggestive, don't you think?

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+    scalar_t* const IP = in.data_ptr<scalar_t>();
+    const scalar_t* const QP = q.data_ptr<scalar_t>();
+
+    auto lt = [](scalar_t a, scalar_t b) {
```
note here about NaN ordering.

Could this be simplified as `(a < b) || _isnan(b)`? Putting the more frequent comparison first may shortcircuit the NaN checks, which is nice. Second, if there's a NaN then isn't it OK to say that A is always less than it in this case, even if A is NaN?

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+        // Check if all quantiles should be set to NaN
+        if (num_nans == reduction_size || (!ignore_nan && num_nans > 0)) {
+          for (int64_t i = 0; i < q_size; ++i, op += num_reductions) {
+            *op = NAN;
```
`*op = std::numeric_limits<scalar_t>::quiet_NaN`

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -188,9 +343,9 @@ void quantile_impl(
-    // compares nans as largest value. If all values are nan, set rank to 0 so that
-    // the quantile computed will be nan.
+    // For nanquantile, compute ranks based on number of non-nan values since
+    // sort compares nans as largest value. If all values are nan, set rank to 0
```
nit: "treats NaNs as larger than any non-NaN values. If all values are NaN, sets rank to zero so the computed quantiles will be NaN, too."

One trick when writing comments is to ensure they describe what the code does ("sets the value to zero") as opposed to giving an order to the reader ("set the value to zero").

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -214,7 +369,7 @@ void quantile_impl(
 
   if (!ignore_nan) {
-    // If there are nan values and nans are not ignored then 
+    // If there are nan values and nans are not ignored then
```
nit: "If NaNs are not ignored (i.e. quantile, not nanquantile is called) then if a NaN is present in a reduced row all the quantiles for that row are also NaN."

mruberry commented on test/test_torch.py:
```diff
@@ -10433,14 +10433,14 @@ def test_quantile(self, device, dtype):
         inputs.extend([[float('nan')], [[float('nan'), float('nan')], [1, 2]]])
         inputs.extend([[[float('nan'), float('nan')], [float('nan'), 2]]])
-        quantiles.extend([0.5, (0., 1.), np.random.rand(10)])
+        quantiles.extend([0.5, [0., 1.], np.random.rand(10)])
```
Interesting - why the change from a tuple to a list here?

mruberry commented on test/test_torch.py:
```diff
@@ -10433,14 +10433,14 @@ def test_quantile(self, device, dtype):
                 a = torch.randn(x, dtype=dtype, device=device)
                 # Make some random elements NaN
-                a.masked_fill_(torch.randint_like(a, 25) == 0, float('nan'))
+                a.masked_fill_(torch.randint_like(a, 20) == 0, float('nan'))
```
Why this change?

mruberry commented:
This is a great start. I made some comment suggestions/requests. I'm curious to look at more performance data and discuss alternative approaches.

heitorschueroff committed:
Update on "(nan)quantile CPU optimization"


This PR replaces a call to sort with a multi-quickselect implementation to partition the data before querying for kth values. When the number of quantiles is small compared to the size of the dimension being reduced, this leads to big performance improvements.

For a single quantile, the previous code was 8 times slower than NumPy. Now it is the 1-2x faster for quantile and up to 8x faster for nanquantile. The benchmark snippet below demonstrates this case.

```
import torch
import numpy as np
import pytest

torch.set_num_threads(1)
torch.set_grad_enabled(False)

x = torch.randn(1000, 1000, dtype=torch.float64)
q = torch.rand(1, dtype=torch.float64)


@pytest.mark.parametrize('op', ['torch', 'numpy'])
def test_quantile(benchmark, op):
    if op == 'torch':
        benchmark(torch.nanquantile, x, q, 1)
    else:
        benchmark(np.nanquantile, x.numpy(), q.numpy(), 1)
```

### Outputs
```
---------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------
Name (time in ms)            Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_quantile[torch]     10.8364 (1.0)      13.4978 (1.0)      11.6419 (1.0)      0.5855 (1.0)      11.7781 (1.0)      0.9811 (1.92)         27;1  85.8970 (1.0)          67           1
test_quantile[numpy]     94.6725 (8.74)     96.9456 (7.18)     95.4756 (8.20)     0.6195 (1.06)     95.3307 (8.09)     0.5101 (1.0)           3;1  10.4739 (0.12)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

[ghstack-poisoned]

heitorschueroff committed:
Update on "(nan)quantile CPU optimization"


This PR replaces a call to sort with a multi-quickselect implementation to partition the data before querying for kth values. When the number of quantiles is small compared to the size of the dimension being reduced, this leads to big performance improvements.

For a single quantile, the previous code was 8 times slower than NumPy. Now it is the 1-2x faster for quantile and up to 8x faster for nanquantile. The benchmark snippet below demonstrates this case.

```
import torch
import numpy as np
import pytest

torch.set_num_threads(1)
torch.set_grad_enabled(False)

x = torch.randn(1000, 1000, dtype=torch.float64)
q = torch.rand(1, dtype=torch.float64)


@pytest.mark.parametrize('op', ['torch', 'numpy'])
def test_quantile(benchmark, op):
    if op == 'torch':
        benchmark(torch.nanquantile, x, q, 1)
    else:
        benchmark(np.nanquantile, x.numpy(), q.numpy(), 1)
```

### Outputs
```
---------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------
Name (time in ms)            Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_quantile[torch]     10.8364 (1.0)      13.4978 (1.0)      11.6419 (1.0)      0.5855 (1.0)      11.7781 (1.0)      0.9811 (1.92)         27;1  85.8970 (1.0)          67           1
test_quantile[numpy]     94.6725 (8.74)     96.9456 (7.18)     95.4756 (8.20)     0.6195 (1.06)     95.3307 (8.09)     0.5101 (1.0)           3;1  10.4739 (0.12)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

[ghstack-poisoned]

heitorschueroff committed:
Update on "(nan)quantile CPU optimization"


This PR replaces a call to sort with a multi-quickselect implementation to partition the data before querying for kth values. When the number of quantiles is small compared to the size of the dimension being reduced, this leads to big performance improvements.

For a single quantile, the previous code was 8 times slower than NumPy. Now it is the 1-2x faster for quantile and up to 8x faster for nanquantile. The benchmark snippet below demonstrates this case.

```
import torch
import numpy as np
import pytest

torch.set_num_threads(1)
torch.set_grad_enabled(False)

x = torch.randn(1000, 1000, dtype=torch.float64)
q = torch.rand(1, dtype=torch.float64)


@pytest.mark.parametrize('op', ['torch', 'numpy'])
def test_quantile(benchmark, op):
    if op == 'torch':
        benchmark(torch.nanquantile, x, q, 1)
    else:
        benchmark(np.nanquantile, x.numpy(), q.numpy(), 1)
```

### Outputs
```
---------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------
Name (time in ms)            Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_quantile[torch]     10.8364 (1.0)      13.4978 (1.0)      11.6419 (1.0)      0.5855 (1.0)      11.7781 (1.0)      0.9811 (1.92)         27;1  85.8970 (1.0)          67           1
test_quantile[numpy]     94.6725 (8.74)     96.9456 (7.18)     95.4756 (8.20)     0.6195 (1.06)     95.3307 (8.09)     0.5101 (1.0)           3;1  10.4739 (0.12)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

[ghstack-poisoned]

heitorschueroff committed:
Update on "(nan)quantile CPU optimization"


This PR replaces a call to sort with a multi-quickselect implementation to partition the data before querying for kth values. When the number of quantiles is small compared to the size of the dimension being reduced, this leads to big performance improvements.

For a single quantile, the previous code was 8 times slower than NumPy. Now it is the 1-2x faster for quantile and up to 8x faster for nanquantile. The benchmark snippet below demonstrates this case.

```
import torch
import numpy as np
import pytest

torch.set_num_threads(1)
torch.set_grad_enabled(False)

x = torch.randn(1000, 1000, dtype=torch.float64)
q = torch.rand(1, dtype=torch.float64)


@pytest.mark.parametrize('op', ['torch', 'numpy'])
def test_quantile(benchmark, op):
    if op == 'torch':
        benchmark(torch.nanquantile, x, q, 1)
    else:
        benchmark(np.nanquantile, x.numpy(), q.numpy(), 1)
```

### Outputs
```
---------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------
Name (time in ms)            Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_quantile[torch]     10.8364 (1.0)      13.4978 (1.0)      11.6419 (1.0)      0.5855 (1.0)      11.7781 (1.0)      0.9811 (1.92)         27;1  85.8970 (1.0)          67           1
test_quantile[numpy]     94.6725 (8.74)     96.9456 (7.18)     95.4756 (8.20)     0.6195 (1.06)     95.3307 (8.09)     0.5101 (1.0)           3;1  10.4739 (0.12)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

[ghstack-poisoned]

heitorschueroff committed:
Update on "(nan)quantile CPU optimization"


This PR replaces a call to sort with a multi-quickselect implementation to partition the data before querying for kth values. When the number of quantiles is small compared to the size of the dimension being reduced, this leads to big performance improvements.

For a single quantile, the previous code was 8 times slower than NumPy. Now it is the 1-2x faster for quantile and up to 8x faster for nanquantile. The benchmark snippet below demonstrates this case.

```
import torch
import numpy as np
import pytest

torch.set_num_threads(1)
torch.set_grad_enabled(False)

x = torch.randn(1000, 1000, dtype=torch.float64)
q = torch.rand(1, dtype=torch.float64)


@pytest.mark.parametrize('op', ['torch', 'numpy'])
def test_quantile(benchmark, op):
    if op == 'torch':
        benchmark(torch.nanquantile, x, q, 1)
    else:
        benchmark(np.nanquantile, x.numpy(), q.numpy(), 1)
```

### Outputs
```
---------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------
Name (time in ms)            Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_quantile[torch]     10.8364 (1.0)      13.4978 (1.0)      11.6419 (1.0)      0.5855 (1.0)      11.7781 (1.0)      0.9811 (1.92)         27;1  85.8970 (1.0)          67           1
test_quantile[numpy]     94.6725 (8.74)     96.9456 (7.18)     95.4756 (8.20)     0.6195 (1.06)     95.3307 (8.09)     0.5101 (1.0)           3;1  10.4739 (0.12)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

[ghstack-poisoned]

heitorschueroff committed:
Update on "(nan)quantile CPU optimization"


This PR replaces a call to sort with a multi-quickselect implementation to partition the data before querying for kth values. When the number of quantiles is small compared to the size of the dimension being reduced, this leads to big performance improvements.

For a single quantile, the previous code was 8 times slower than NumPy. Now it is the 1-2x faster for quantile and up to 8x faster for nanquantile. The benchmark snippet below demonstrates this case.

```
import torch
import numpy as np
import pytest

torch.set_num_threads(1)
torch.set_grad_enabled(False)

x = torch.randn(1000, 1000, dtype=torch.float64)
q = torch.rand(1, dtype=torch.float64)


@pytest.mark.parametrize('op', ['torch', 'numpy'])
def test_quantile(benchmark, op):
    if op == 'torch':
        benchmark(torch.nanquantile, x, q, 1)
    else:
        benchmark(np.nanquantile, x.numpy(), q.numpy(), 1)
```

### Outputs
```
---------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------
Name (time in ms)            Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_quantile[torch]     10.8364 (1.0)      13.4978 (1.0)      11.6419 (1.0)      0.5855 (1.0)      11.7781 (1.0)      0.9811 (1.92)         27;1  85.8970 (1.0)          67           1
test_quantile[numpy]     94.6725 (8.74)     96.9456 (7.18)     95.4756 (8.20)     0.6195 (1.06)     95.3307 (8.09)     0.5101 (1.0)           3;1  10.4739 (0.12)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

[ghstack-poisoned]
